### PR TITLE
AGI: [WIP, RFC] Optionally enable copy protection in Software Farm's Gold Rush

### PR DESCRIPTION
**Pull request by eriktorbjorn**
This deceptively simple patch should not be merged without a thorough review.

The Software Farm re-release of Gold Rush! doesn't show any copy protection screen. It also doesn't include the necessary documentation for it. Perhaps the MacNeills only kept the rights to the game, not that book?

But the copy protection screen is still there in the data files, so perhaps we should add an option to enable it just for fun? I had never even seen it before today!

Some notes:

- The way it works is that it changes a "new room" instruction to go to the copy protection screen, if it's called from one particular logic. I don't know much about AGI, so I don't know if this is the appropriate way or if we can do better.
- The wording of the setting is a bit different than the recently worded copy protection setting in the SCUMM engine. This is deliberate. In SCUMM, we enable copy protection that ScummVM bypasses. Here we enable copy protection that the authors of the game has disabled. Perhaps it can be reworded to cover both these cases?
- This is the only release of Gold Rush that I own. I hope the change doesn't negatively affect any other releases. Come to think of it, are there other releases that disabled the copy protection or is it just this one?

### CODE REVIEW

**Comment by sluicebox** [comment]
Who would want this? I'm kind of baffled, and against it. This is Jeff Goldblum Jurassic Park Meme territory.

Real Talk: if this weren't @eriktorbjorn , whom I know to be sincere and thoughtful, I would close this PR as trolling. Adding copy protection to games that didn't have them is not something I thought I'd have to defend against. (Although it's a funny idea!)

The Gold Rush authors removed Sierra's copy protection once they could distribute it themselves in 1998. As you've pointed out, the documentation no longer matched (reason enough to reject this!), but I'm certain they removed it because it's the worst part of the software that no one wanted except Sierra. They advertise it right now on their website as "DRM Free!" Why would we offer to degrade this? If they wanted copy protection they would have just updated the game scripts in 1998. In addition to this being a feature authors and players don't want, I think it's disrespectful to resurrect this pain just because they forgot to delete the dead resources. If the Sierra logo were still buried in there, we wouldn't offer to replace The Software Farm logo just because we can.

Anyone who wants Gold Rush copy protection (should such a person exist) can get the very obtainable Sierra version. They probably already have it. That's the popular one. Gold Rush 3.0 is the rare version, even though it's purchasable. As someone who has spent a lot of time backporting Sierra bugfixes to old versions, I'm stunned at the idea of infecting a newer version with the worst part of an old one. Which again, players won't be able to pass without a 2.0 manual. [ BTW; the scene in the copy protection is part of the real game, so really it's a spoiler. ]

The minimum bar for an option should be that someone would use it and that it's good. Even ignoring that it's only available to 3.0 owners, I don't this meets that bar. The idea of also forcing all our translators to translate the user interface does not go over well with me; it feels like just making work.

I feel silly saying this, but we will not be enabling SCI copy protection in CD versions just because the unused floppy version scripts are floating around on the discs.

Earlier you suggested we alter KQ4 copy protection because a CD collection made a few documentation answers incorrect. (Something we couldn't do anything about, since that's ambiguous.) Now you're suggesting we introduce copy protection for a version that will never match its documentation. What's going on here? =)

**Comment by eriktorbjorn** [comment]
> They advertise it right now on their website as "DRM Free!"

I must be looking in the wrong places, because I couldn't find any mention of that on their web page. (I was looking at www.softwarefarm.com which, unfortunately, still seems to be under construction.) I'm not being facetious, "gold rush" is one of those search terms that bring up a lot of false positives.

> If the Sierra logo were still buried in there, we wouldn't offer to replace The Software Farm logo just because we can.

Funny you should mention those, because this all started with me looking for them.

Only to satisfy my curiosity, mind you. Looking at YouTube videos and screenshots, there seemed to be two different ones. One that seemed vaguely familiar ("the MacNeill bros. MacNeill Software Development MacNeill Computer Company and SIERRA are proud to present") because I saw the game back in the eighties. One that just said "and SIERRA", as if something had been cut out.

I couldn't find them, but I did stumble over the copy protection quiz which I had never, ever seen given the **un**availability of any other versions.

As for who'd want it... I think it's a nice piece of AGI graphics, rarely seen outside of YouTube. While looking for it, I found a few commenters praising its creativity, and talking about how it scared them as kids (which I'm all for!), and someone lamenting that all he could find was cracked disk images with the copy protection quiz removed.

> I feel silly saying this, but we will not be enabling SCI copy protection in CD versions just because the unused floppy version scripts are floating around on the discs.

Aw, now you've made me curious about those! I have a love/hate relation with copy protection. I dislike having to look it up, but one of my favorite LucasArts jokes comes from failing at it in Indiana Jones and the Last Crusade. It's a love/hate relationship. Or at least a somewhat-dislike/kind-of-enjoy one.

> Earlier you suggested we alter KQ4 copy protection because a CD collection made a few documentation answers incorrect. (Something we couldn't do anything about, since that's ambiguous.)

I think that's a different situation. KQ4 has an unskippable "look up the word in the manual" check (already one of the laziest forms of copy protection known to man), with a 3-4% chance (I think) of asking you for something not included in the manual it's been most commonly sold with all the way back since Sierra started making collections.

I've been tempted to suggest the same thing for Future Wars, where the manual I have crops several of the paint spots it asks for out of existence. But then I remember that anyone playing that game is already a glutton for punishment. :-)

There have been other cases where they have left out copy protection information. Most egregiously, the first Police Quest collection I bought (new, not second hand) didn't have any documentation at all except for installation and registration instructions, and a note from Daryl Gates thanking me for buying his SWAT collection. But I assume that was an aberration.

But I welcome the discussion! That's why I made this a pull request and marked it "RFC". If people think it's a stupid idea, then by all means it should be rejected. It's happened before, it will almost certainly happen again, but I've never regretted trying because I had fun doing it.

**Comment by einstein95** [comment]
I've had a look at a bunch of versions (included ones detected as pirated), and since the v2.01 versions have working copy protection in ScummVM despite v3.0 bypassing it, I'd change the patch to be
```cpp
if (getGameID() == GID_GOLDRUSH && _game.curLogicNr == 129) {
    newRoomNr = ConfMan.getBool("copy_protection") ? 125 : 73;
}
```
as this both enables people to see the original copy protection in v3.0 (where one of the questions had a grammar correction ("A tailing wheel 68 feet in diameter was built in 1912 to take tailings from this mine, and dump **it** into a dam behind Jackson." -> "A tailing wheel 68 feet in diameter was built in 1912 to take tailings from this mine, and dump **them** into a dam behind Jackson.") as well as bypassing the copy protection in the older releases.

This is basically the same way as how the copy protection was "removed" in 3.0:
```diff
   if (Restarting) {
     animationInterval = 2;
     new.room(rm.MainSt);
   } else {
-    // just starting - run the copy protection sequence
-    new.room(lgc.InitQuiz);  // 125
+    // just starting - skip to title
+    new.room(rm.Title);  // 73
   }
```

> One that just said "and SIERRA", as if something had been cut out.
This is from a pirated version that replaces the pointer to room 125 to room 73, which causes this issue.

**Comment by eriktorbjorn** [comment]
> as this both enables people to see the original copy protection in v3.0 (where one of the questions had a grammar correction ("A tailing wheel 68 feet in diameter was built in 1912 to take tailings from this mine, and dump it into a dam behind Jackson." -> "A tailing wheel 68 feet in diameter was built in 1912 to take tailings from this mine, and dump them into a dam behind Jackson.") as well as bypassing the copy protection in the older releases.

That would solve the translation issue , since the AGI setting could use the exact string as the SCUMM setting. The reasons I didn't do that was because because I wasn't sure we should bypass the copy protection of other versions or not, and because the Software Farm version is the only one I have access to. My reasoning was that we bypass the copy protection when we there is an official release that does, and we can't tell it apart from the other ones. Here we can.

But there is some precedent for bypassing the copy protection anyway, e.g. the Macintosh version of The Secret of Monkey Island. I own the two versions from the _LucasArts Mac CD Game Pack_ and _LucasArts Mac CD Game Pack II_. The adventure games on the first one (Monkey Island 1, Last Crusade, and Loom) all have copy protection, while the ones in the second one (Monkey Island 1, Monkey Island 2, and Fate of Atlantis) have those parts of the scripts removed.

I suggested that we shouldn't bypass the copy protection in Monkey Island 1, since it's already bypassed in that version so we should leave the others alone, but I was voted down. The Gold Rush situation seems pretty similar to me.

See https://github.com/scummvm/scummvm/pull/5413 for details.

> > One that just said "and SIERRA", as if something had been cut out.
> 
> This is from a pirated version that replaces the pointer to room 125 to room 73, which causes this issue.

I suspected as much. Thanks for confirming it! What confused me was that one screen was a nighttime scene, and the other was in daylight, but I guess it's showing a completely unrelated background. Of course, now that I asked I happened to stumble over this:

https://youtu.be/JtioyqqDCGI?t=1091

I have to say that I really like the "The MacNeill bros."  version, which is why I went looking for it in the first place. I realize now that the "Software Farm" one still uses the same background, I just don't think it looks _quite_ as nice. Of course I can see why they did it, and I'm actually quite impressed that they did. The most I've seen/noticed the Sierra re-releases do is to cross out a few phone numbers.

**Comment by eriktorbjorn** [comment]
> This is from a pirated version that replaces the pointer to room 125 to room 73, which causes this issue.

By the way, does this mean that if ScummVM bypasses the copy protection it will also break the title screen? Because that would be a shame.

**Comment by bluegr** [comment]
The whole point of a copy protection screen back then was to verify the authenticity of the copy, using physical medium (such as manuals, code wheels etc). Since this version did not include the physical medium, i.e. in this case, the manual, what is the point of enabling the copy protection screen?

I can only guess that players will start the game with the copy protection checkbox checked, view the copy protection screen and  either:
- exit and restart without the copy protection (since they can't find the answers)
- search for the copy protection answers online

Again, what is the point of this change, apart from frustrating players, or forcing them to search online for obscure answers to copy protection questions, in a manual that they don't own?

**Comment by eriktorbjorn** [comment]
> Again, what is the point of this change,

If it had been a King's Quest 4 style thing, I would not have seen much point in it. But as I said before, I think this is an interesting piece of animated AGI artwork that, given the unavailability of all other versions, I would have no other way of ever seeing for myself.

goldrush

There have been other cases where we've enabled artwork that wasn't originally seen, or was only present in other versions of a game. I don't see this as being _that_ different.

> (since they can't find the answers)

The questions seem to be largely based on historical facts, so you could argue that even without the exact manual they're arguably still educational. Fitting for a game that advertises that it "Teaches historical and geographical facts."

**Comment by bluegr** [comment]
> > Again, what is the point of this change,
> 
> If it had been a King's Quest 4 style thing, I would not have seen much point in it. But as I said before, I think this is an interesting piece of animated AGI artwork that, given the unavailability of all other versions, I would have no other way of ever seeing for myself.

Viewing a piece of artwork is different from enabling a copy protection screen. All such screens may have their own artwork (e.g. the pirates in the dial-a-pirare copy protection screen in MI1), but this is by no means a reason to enable them.

> There have been other cases where we've enabled artwork that wasn't originally seen, or was only present in other versions of a game. I don't see this as being _that_ different.

Of course it is different to view an artwork (which may be done with a resource viewer) than actually enable a room that enforces a copy protection scheme.


> > (since they can't find the answers)
> 
> The questions seem to be largely based on historical facts, so you could argue that even without the exact manual they're still educational. Fitting for a game that advertises that it "Teaches historical and geographical facts."

It seems you are really trying to convince me that a copy protection screen is meant to be "fun". It's not, it's there to protect the game from being copied :)

Historical facts in this case are targeted at native US citizens. Also, a game that asks for historical facts is meant to be fun and teach its player base of such facts, not hang their hero and exit the game in the case of a wrong answer!

**Comment by eriktorbjorn** [comment]
> All such screens may have their own artwork (e.g. the pirates in the dial-a-pirare copy protection screen in MI1), but this is by no means a reason to enable them.

We do offer the option to enable the dial-a-pirate screen though, and a couple of others. Sometimes I _do_ think it's fun to do that.

**Comment by raziel-** [comment]
if i may throw my two cents in the ring?

TL;DR
Show the scene, but skip the questioning?

Long version:

i own the amiga version of Gold Rush *with* the copy protection and always thought it was a unique way to deal with wrong answers, instead of simply showing a "you failed, no game for you" window or text.

Also, the artwork of the hanging scene let me try again, because i *wanted* to play the game afterwards.

The facts in the "history" book were also interesting, even for a non-native US citizen...normally people play games because of the story, won't they?

IF it is technical possible, why not show the whole scene, on removed cooy protection versions, but tell the player about it beforehand and simply play it and move on into the game?

like, a window telling that this was the copy protection screens and when one would be thrown out of the game, if answered wrong?

Either in-game or when hovering over the option?

Maybe even an option for other games who skipped this content?

**Comment by eriktorbjorn** [comment]
> IF it is technical possible, why not show the whole scene, on removed cooy protection versions, but tell the player about it beforehand and simply play it and move on into the game?

I don't know... I thought the rule was that we skip the copy protection if we can justify it, but we don't change the way it works when it's not skipped. See for instance https://bugs.scummvm.org/ticket/14595

Unless the copy protection itself has been buggy, which apparently was the case with the fingerprint check in The Colonel's Bequest. But in Gold Rush, the copy protection seems to be working just as intended.

The the options I can see are:

1. Leave things the way they are.
2. Make it possible to enable the copy protection in the Software Farm version, but leave it they way it is in other versions.
3. Disable the copy protection in all versions. Based on how we've handled that in the past, it should obey the "copy_protection" config key, in which case the Software Farm might gain the ability to enable the copy protection too.
4. Disable the copy protection in all versions, but _don't_ allow the copy protection to be enabled in the Software Farm version.

So far, most seem to be arguing for option 1?

Whichever option we take, I don't think we should assume that those who are nostalgic for, or curious about, the copy protection have access to one of those releases. (I don't.) Or know how to use an AGI resource viewer. (I don't.)

**Comment by tsoliman** [comment]
It seems the proposal defaults the copy protection to off. This is a good thing. It'd be there as a novelty option for those who want it.
I propose a fifth option: disable copy protection in all versions by default, and allow them all to be enabled as a novelty using this new engine option (including the Software Farm version). Seems to be the most user friendly option IMHO.

**Comment by eriktorbjorn** [comment]
> It seems the proposal defaults the copy protection to off. This is a good thing. It'd be there as a novelty option for those who want it.

Right. The intention was never to enable the copy protection by default, because the re-release doesn't include the "California Gold" booklet. It's credited to "Phyllis and Lou Zauner". A quick search on Amazon turns up other similar (?) books by Phyllis Zauner. In fact "California Gold" is listed as being published in 1980, so maybe Sierra licensed it for use with the game (which was released in 1988)?

The idea was to give people like me (who had never seen the copy protection quiz), or people who had seen it and were nostalgic for it - judging by https://www.youtube.com/watch?v=kN7JgX1rfjs there may be a few - but who no longer have the old version the option to revisit it.

> I propose a fifth option: disable copy protection in all versions by default, and allow them all to be enabled as a novelty using this new engine option (including the Software Farm version). Seems to be the most user friendly option IMHO.
> 
> EDIT: this might be option 3 actually. I couldn't tell.

Yes, that's what I meant by option 3. But I didn't want to touch the old releases because a) I didn't know if I should, and b) I don't own any of the old releases (since I have no floppy drive, it seems unlikely that I ever will) so I wouldn't be able to test it anyway.

**Comment by lephilousophe** [comment]
I think Option 3 is the way to go.
Everyone get a version without copy protection and purists can enable it either by the `--copy-protection` either by the settings screen.
There is already an option like this in numerous engines (AGOS, Dreamweb, Gob, Lure, MADS, MM, SAGA, SCUMM and Voyeur).

**Comment by eriktorbjorn** [comment]
> I think Option 3 is the way to go.

I'm not going to oppose that, I'm just not sure what our policy is when a later release removes copy protection, but we don't have to do anything to skip the copy protection in that release. Because I can think of one example where we do bypass the copy protection in earlier releases too, and one where we don't.

The one where we do is the Macintosh version of The Secret of Monkey Island. There exists a release (The LucasArts Mac CD Pack II) where the part of the script that starts the copy protection has been cut out, and we use that as an excuse to skip it in all Mac versions of the game. See the aforementioned https://github.com/scummvm/scummvm/pull/5413

The same release also strips out the copy protection from Monkey Island 2 and Indiana Jones and the Last Crusade. Even with the --copy-protection option, you won't get the copy protection... except in Monkey Island 2, where the missing script has been re-inserted to enable easy mode. (The copy protection is then skipped with a Mac-specific boot param.)

The one where we don't is Future Wars. The floppy version asks you to identify the color of a randomly selected paint spot in a picture. This was removed from the CD version, but ScummVM does not bypass the protection in the floppy version. This copy protection is very annoying if, like me, you own the Delphine Software Classic Collection. Not only are the colors a bit washed out, but several of the paint spots have also been cropped out of the image. This image shows the possible spots (with redacted colors), and the approximate part of it that's included in that manual.

fw-cp-fail

"We have perfected a copy protection system which we hope will not cause you too much inconvenience" my foot!

@sev- made the call on The Secret of Monkey Island, so I'd be interested to hear his opinion here.

**Comment by eriktorbjorn** [comment]
I'll try to update this pull request later (I didn't have the time yesterday or this morning), though I'm going to need a bit of help testing it afterwards.

**Comment by eriktorbjorn** [comment]
On second thought, with the conflicts and whatnot it's probably easier to start over. (Especially since I made a patch based on the current master while experimenting earlier today.) Closing this one
